Remotion/Core/UnitTests/Context/AsyncLocalStorageProviderTest.cs

```diff
@@ -0,0 +1,161 @@
+    }
+
+    [Test]
+    public void GetData_WithoutValue ()
+    {
+      Assert.That (_provider.GetData ("Foo"), Is.Null);
+    }
+
+    [Test]
+    public void SetData ()
+    {
+      _provider.SetData ("Foo", 45);
+      Assert.That (_provider.GetData ("Foo"), Is.EqualTo (45));
+    }
+
+    [Test]
+    public void SetData_Null ()
+    {
+      _provider.SetData ("Foo", 45);
+      _provider.SetData ("Foo", null);
+      Assert.That (_provider.GetData ("Foo"), Is.Null);
+    }
+
+    [Test]
+    public void FreeData ()
+    {
+      _provider.SetData ("Foo", 45);
+      _provider.FreeData ("Foo");
+      Assert.That (_provider.GetData ("Foo"), Is.Null);
+    }
+
+    [Test]
+    public void FreeData_WithoutValue ()
+    {
+      _provider.FreeData ("Foo");
+      Assert.That (_provider.GetData ("Foo"), Is.Null);
+    }
+
+    [Test]
+    public void SetData_FromOtherThread_NotAccessible ()
+    {
+      const string foo = "Foo";
+
+      var t = new Thread (() => _provider.SetData (foo, "1"));
+      t.Start();
+      t.Join();
+
+      Assert.That (_provider.GetData (foo), Is.Null);
+    }
+
+    [Test]
+    public void SetData_FromOtherThread_DoesNotOverride ()
+    {
+      const string foo = "Foo";
+      _provider.SetData (foo, "1");
+
+      var t = new Thread (() => _provider.SetData (foo, "2"));
+      t.Start();
+      t.Join();
+
+      Assert.That (_provider.GetData (foo), Is.EqualTo("1"));
+    }
+
+    [Test]
+    public void SetData_OnThreadPoolThread_IsResetForNextWorkItem ()
+    {
+      const string foo = "Foo";
+
+      object result = string.Empty;
+      int threadId1 = -1;
+      int threadId2 = -2;
+      ThreadPool.GetMaxThreads (out var workerThreads, out var completionPortThreads);
+      try
+      {
+        ThreadPool.SetMaxThreads (1, 1);
+        using var resetEvent = new AutoResetEvent (false);
+
+        ThreadPool.QueueUserWorkItem (
+            _ =>
+            {
+              threadId1 = Thread.CurrentThread.ManagedThreadId;
+              _provider.SetData (foo, "1");
+              resetEvent.Set();
+            });
+        resetEvent.WaitOne();
+        ThreadPool.QueueUserWorkItem (
+            _ =>
+            {
+              threadId2 = Thread.CurrentThread.ManagedThreadId;
+              result = _provider.GetData (foo);
+              resetEvent.Set();
+            });
+        resetEvent.WaitOne();
+      }
+      finally
+      {
+        ThreadPool.SetMaxThreads (workerThreads, completionPortThreads);
+      }
+
+      Assert.That (result, Is.Null);
+      Assert.That (threadId1, Is.EqualTo(threadId2));
+    }
+
+#if NETFRAMEWORK && DEBUG
+
+    [Test]
+    public void SetData_WithClassImplementingILogicalThreadAffinative_ThrowsPlatformNotSupportedException ()
+    {
+      var data = new ClassImplementingILogicalThreadAffinative();
+      Assert.That (
+          () => _provider.SetData ("_", data),
+          Throws.InstanceOf<PlatformNotSupportedException>()
```

Comment:
As you like: if the message fits in the same line as the exception, I don't do a linebreak. I don't mind if it's there, it's just not something I typically do myself.